java-frontend/src/main/java/org/sonar/java/ExecutionTimeReport.java

```diff
@@ -0,0 +1,107 @@
+import org.sonar.api.utils.log.Loggers;
+
+public class ExecutionTimeReport {
+  private static final Logger LOG = Loggers.get(ExecutionTimeReport.class);
+
+  private static final int MAX_REPORTED_FILES = 10;
+  private static final long MIN_REPORTED_ANALYSIS_TIME_MS = 1000;
+  private static final long MIN_TOTAL_ANALYSIS_TIME_TO_REPORT_MS = 20_000;
+
+  private static class ExecutionTime {
+    private final String file;
+    private final long analysisTime;
+
+    public ExecutionTime(String file, long analysisTime) {
+      this.file = file;
+      this.analysisTime = analysisTime;
+    }
+  }
+
+  private static final Comparator<ExecutionTime> ORDER_BY_ANALYSIS_TIME_DESCENDING_AND_FILE_ASCENDING = (a, b) -> {
+    int compare = Long.compare(b.analysisTime, a.analysisTime);
+    return compare != 0 ? compare : a.file.compareTo(b.file);
+  };
+
+  private final LinkedList<ExecutionTime> recordedOrderedExecutionTime = new LinkedList<>();
+  private long minRecordedOrderedExecutionTime = MIN_REPORTED_ANALYSIS_TIME_MS;
+
+  private final Clock clock;
+  private final long analysisStartTimeMS;
+  private String currentFile;
+  private long currentFileStartTimeMS;
+
+  public ExecutionTimeReport(Clock clock) {
+    this.clock = clock;
+    analysisStartTimeMS = clock.millis();
+  }
+
+  public void start(String currentFile) {
+    this.currentFile = currentFile;
+    currentFileStartTimeMS = clock.millis();
+  }
+
+  public void end() {
+    long currentAnalysisTime = clock.millis() - currentFileStartTimeMS;
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Analysis time of " + currentFile + " (" + currentAnalysisTime + "ms)");
+    } else if (currentAnalysisTime >= MIN_REPORTED_ANALYSIS_TIME_MS && LOG.isDebugEnabled()) {
+      LOG.debug("Analysis time of " + currentFile + " (" + currentAnalysisTime + "ms)");
+    }
+    if (currentAnalysisTime >= minRecordedOrderedExecutionTime) {
+      recordedOrderedExecutionTime.add(new ExecutionTime(currentFile, currentAnalysisTime));
+      recordedOrderedExecutionTime.sort(ORDER_BY_ANALYSIS_TIME_DESCENDING_AND_FILE_ASCENDING);
+      if (recordedOrderedExecutionTime.size() > MAX_REPORTED_FILES) {
+        recordedOrderedExecutionTime.removeLast();
+        minRecordedOrderedExecutionTime = recordedOrderedExecutionTime.stream()
+          .mapToLong(e -> e.analysisTime)
+          .min()
+          .orElse(MIN_REPORTED_ANALYSIS_TIME_MS);
+      }
+    }
+    this.currentFile = null;
+  }
+
+  public void report() {
+    if (currentFile != null) {
+      end();
+    }
+    long analysisEndTimeMS = clock.millis() - analysisStartTimeMS;
+    if (analysisEndTimeMS >= MIN_TOTAL_ANALYSIS_TIME_TO_REPORT_MS && !recordedOrderedExecutionTime.isEmpty()) {
+      LOG.info("Slowest analyzed files: " + toString());
+    }
+  }
+
+  @Override
+  public String toString() {
+    return recordedOrderedExecutionTime.stream()
+      .map(e -> e.file + " (" + e.analysisTime + "ms)")
+      .collect(Collectors.joining(", "));
```

Comment:
I really think it is OK to have them on multiple lines, especially since we are displaying the path of each input file (it take place) and only 10 files. I prefer have 10 explicit lines rather than one big chunk of text,

Reply:
ok

Reply:
The new logs look like this:
```
[INFO] 1370/1370 source files have been analyzed
[INFO] Slowest analyzed files:
    src/main/java/org/sonar/server/issue/IssueQueryFactory.java (2143ms)
    src/main/java/org/sonar/ce/CeModule.java (1363ms)
    src/main/java/org/sonar/server/rule/RegisterRules.java (1288ms)
[INFO] Java Main Files AST scan (done) | time=87830ms
```